src/net/neighbors/db.rs

```diff
@@ -0,0 +1,652 @@
+            }
+        }
+
+        tx.commit()?;
+
+        // neighbor was new
+        Ok((true, neighbor_from_handshake))
+    }
+
+    /// implements get_initial_walk_neighbors
+    fn inner_get_initial_walk_neighbors(
+        network: &PeerNetwork,
+        ibd: bool,
+    ) -> Result<Vec<Neighbor>, net_error> {
+        let allowed_peers = if ibd {
+            // only get bootstrap peers (will be randomized)
+            PeerDB::get_bootstrap_peers(
+                &network.peerdb_conn(),
+                network.get_local_peer().network_id,
+            )?
+        } else {
+            // can be any peer marked 'always-allowed' (will be randomized)
+            PeerDB::get_always_allowed_peers(
+                network.peerdb_conn(),
+                network.get_local_peer().network_id,
+            )?
+        };
+        Ok(allowed_peers)
+    }
+
+    /// implements check_neighbor_denied
+    fn inner_check_neighbor_denied(
+        network: &PeerNetwork,
+        nk: &NeighborKey,
+    ) -> Result<(), net_error> {
+        // don't proceed if denied
+        if PeerDB::is_peer_denied(
+            &network.peerdb_conn(),
+            nk.network_id,
+            &nk.addrbytes,
+            nk.port,
+        )? {
+            debug!(
+                "{:?}: neighbor {:?} is denied",
+                network.get_local_peer(),
+                nk
+            );
+            return Err(net_error::Denied);
+        }
+        Ok(())
+    }
+
+    /// implements replace_neighbors
+    fn inner_replace_neighbors(
+        network: &mut PeerNetwork,
+        replacements: &NeighborReplacements,
+        result: &mut NeighborWalkResult,
+    ) -> Result<(), net_error> {
+        let network_id = network.bound_neighbor_key().network_id;
+        let local_peer_str = format!("{:?}", network.get_local_peer());
+
+        let mut tx = network.peerdb_tx_begin()?;
+        for (replaceable_naddr, slot) in replacements.iter_slots() {
+            let replacement = match replacements.get_neighbor(replaceable_naddr) {
+                Some(n) => n,
+                None => {
+                    continue;
+                }
+            };
+
+            let replaced_opt = PeerDB::get_peer_at(&mut tx, network_id, *slot)?;
+            if let Some(replaced) = replaced_opt {
+                if PeerDB::is_address_denied(&mut tx, &replacement.addr.addrbytes)? {
+                    debug!(
+                        "{:?}: Will not replace {:?} with {:?} -- is denied",
+                        local_peer_str, &replaced.addr, &replacement.addr
+                    );
+                } else {
+                    debug!(
+                        "{:?}: Replace {:?} with {:?}",
+                        local_peer_str, &replaced.addr, &replacement.addr
+                    );
+
+                    PeerDB::insert_or_replace_peer(&mut tx, &replacement, *slot)?;
+                    result.add_replaced(replaced.addr.clone());
+                }
```

Comment:
Does this _only_ insert `replacement` if the slot was occupied? What if the slot was vacant? Shouldn't the new neighbor still be inserted?